flink-core/src/main/java/org/apache/flink/api/common/operators/util/PoissonSampler.java

```diff
@@ -0,0 +1,109 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.api.common.operators.util;
+
+import com.google.common.base.Preconditions;
+import org.apache.commons.math3.distribution.PoissonDistribution;
+
+import java.util.Iterator;
+
+/**
+ * A sampler implementation based on Poisson Distribution. While sampling elements with fraction and replacement,
+ * the selected number of each element follows a given poisson distribution, so we could use poisson
+ * distribution to generate random variables for sample.
+ *
+ * @param <T> The type of sample.
+ * @see <a href="https://en.wikipedia.org/wiki/Poisson_distribution">https://en.wikipedia.org/wiki/Poisson_distribution</a>
+ */
+public class PoissonSampler<T> extends RandomSampler<T> {
+	
+	private PoissonDistribution poissonDistribution;
+	private final double fraction;
+	
+	/**
+	 * Create a poisson sampler which would sample elements with replacement.
+	 *
+	 * @param fraction The expected count of each element.
+	 * @param seed     Random number generator seed for internal PoissonDistribution.
+	 */
+	public PoissonSampler(double fraction, long seed) {
+		Preconditions.checkArgument(fraction >= 0, "fraction should be positive.");
+		this.fraction = fraction;
+		if (this.fraction > 0) {
+			this.poissonDistribution = new PoissonDistribution(fraction);
+			this.poissonDistribution.reseedRandomGenerator(seed);
+		}
+	}
+	
+	/**
+	 * Create a poisson sampler which would sample elements with replacement.
+	 *
+	 * @param fraction The expected count of each element.
+	 */
+	public PoissonSampler(double fraction) {
+		Preconditions.checkArgument(fraction >= 0, "fraction should be non-negative.");
+		this.fraction = fraction;
+		if (this.fraction > 0) {
+			this.poissonDistribution = new PoissonDistribution(fraction);
+		}
+	}
+	
+	/**
+	 * Sample the input elements, for each input element, generate its count with poisson distribution random variables generation.
+	 *
+	 * @param input Elements to be sampled.
+	 * @return The sampled result which is lazy computed upon input elements.
+	 */
+	@Override
+	public Iterator<T> sample(final Iterator<T> input) {
+		if (fraction == 0) {
+			return EMPTY_ITERABLE;
+		}
+		
+		return new SampledIterator<T>() {
+			T currentElement;
+			int currentCount = 0;
+			
+			@Override
+			public boolean hasNext() {
+				if (currentElement == null || currentCount == 0) {
+					while (input.hasNext()) {
+						currentElement = input.next();
+						currentCount = poissonDistribution.sample();
+						if (currentCount > 0) {
+							return true;
+						}
+					}
+					return false;
+				}
+				return true;
+			}
+			
+			@Override
+			public T next() {
+				T result = currentElement;
+				if (currentCount == 0) {
+					currentElement = null;
+					return null;
+				}
```

Comment:
Can we put the following statements in an else branch and remove the `T result = currentElement;`? I think this makes the control flow slightly more explicit.